src/Avalonia.X11/X11Window.cs

```diff
@@ -195,6 +197,42 @@ public X11Window(AvaloniaX11Platform platform, IWindowImpl? popupParent, bool ov
                 new X11FramebufferSurface(_x11.DeferredDisplay, _renderHandle, 
                    depth, _platform.Options.UseRetainedFramebuffer ?? false)
             };
+
+            if (_platform.Options.UseXShmFramebuffer is true)
+            {
+                TryInsertX11ShmFramebufferSurface();
+
+                void TryInsertX11ShmFramebufferSurface()
+                {
+                    if (depth != 32)
+                    {
+                        // If the depth is not 32, we should do some conversion to make the XShmPutImage work. But the conversion is slowly, so we should not use XShmPutImage when the depth is not 32.
+                        return;
+                    }
+
+                    var status = XShmQueryExtension(_x11.DeferredDisplay);
+                    if (status == 0)
+                    {
+                        // XShmQueryExtension failed
+                        return;
+                    }
+
+                    status = XShmQueryVersion(_x11.DeferredDisplay, out var major, out var minor, out var pixmaps);
+                    
+                    if (status == 0)
+                    {
+                        // XShmQueryExtension failed
+                        return;
+                    }
+
+                    // The reason for using `_x11.Display` instead of `_x11.DeferredDisplay` is that XSHM requires pushing to a Display that can receive Events, in order to obtain the XShmCompletionEvent when rendering is complete.
```

Comment:
DeferredDisplay should be used for rendering since compositor runs on a separate thread. You can get events from that connection too as previously discussed in https://github.com/AvaloniaUI/Avalonia/discussions/16690#discussioncomment-10378881 (see DeferredDisplayEvents class)

Reply:
@kekekeks The DeferredDisplay do not call the `XSelectInput`, and I do not know how to recevice the XShmCompletionEvent from DeferredDisplay. Thank you.